### PR TITLE
Pre-CRAN Release

### PR DESCRIPTION
**Pull request by russellpierce**

### CODE REVIEW

**Comment by russellpierce** [comment]
## [Current coverage](https://codecov.io/gh/zapier/awsjavasdk/pull/1?src=pr) is 100% (diff: 100%)


> No coverage report found for **master** at 3a501a0.

> Powered by [Codecov](https://codecov.io?src=pr). Last update [3a501a0...105f7b1](https://codecov.io/gh/zapier/awsjavasdk/compare/3a501a014eb26dce30f77c54ceb40f68b0c05740...105f7b1eebc3d5e9cb091351d1b584f23eff0972?src=pr)

File: R/awsjavasdk-internal.R [hunk 1]
```diff
@@ -17,8 +17,9 @@
 .app_dir <- rappdirs::app_dir("awsjavasdk", version = packageVersion("awsjavasdk"))
 
 .onAttach <- function(libname, pkgname) {
+  set_sdk_file_root()
   if (!aws_sdk_present()) {
-    packageStartupMessage("awsjavasdk did not detect its own copy of the AWS SDK; you must execute install_aws_sdk() in order to download the SDK and for the package to work as expected")
+    packageStartupMessage(paste0("awsjavasdk did not detect its own copy of the AWS SDK. You must execute install_aws_sdk() in order to download the SDK and for the package to work as expected. If you've previously installed to a non-standard location, call set_sdk_file_root({{your alt_install root}}, confirm = TRUE) to point this package to that location and then run awsjavasdk::load_sdk()"))
```
Comment (daroczig) [inline R/awsjavasdk-internal.R:22]: @gaborcsardi and @jeroenooms do you know any better way of installing a bunch of jar files in an R package? I'm having the same struggle with CRAN policies as Russel here: the AWS jar files are pretty big (~5 - 40Mb depending on how many AWS services you want to include), so cannot put on CRAN.

I was thinking about creating a drat repo on GH, but installation from there also requires extra work from the user at installation time. I'm thinking about a way to automatically download the libs when `R CMD build` runs instead of asking the user to download/install the jars or to provide a helper function so that they have to only confirm that they are OK with this.

In short, any nicer (from the user's perspective) way to handle this? Eg is it possible/legal to use a Makefile to download the jars and add to the package? Or should I just get along with this approach so that users/scripts have to do others things as well besides `install.packages` eg to install my KMS and Kinesis R packages?

Comment (gaborcsardi) [reply R/awsjavasdk-internal.R:22]: I would put them in another package. A lot of people do that already: https://www.r-pkg.org/search.html?q=jars

Comment (daroczig) [reply R/awsjavasdk-internal.R:22]: Thanks, @gaborcsardi -- that was my intention as well and I was just about to come up with an `AWR` (AWS jars for R) data package for this purpose, but having found this repository I *think* CRAN does not want to host such package.

If that's not the case, I'm not sure why @drknexus came up with this less natural solution.

Comment (russellpierce) [reply R/awsjavasdk-internal.R:22]: @daroczig I can't speak as to whether CRAN wants to host such a package - so take heart!  However, space certainly isn't a concern per se (if updated rarely), and others have hosted Apache Licensed JARs there before.  So, you should be fine to pursue that route in theory.  The rub is in practice.  How up-to-date do you want the jars to be?

AWS releases patches at an astonishing pace (https://aws.amazon.com/releasenotes/).  I don't expect that CRAN wouldn't tolerate a non-huge package publishing at that rate, much less a large one (because they state the advantage of a large jar-only packages is that they can be updated rarely).  Maybe you have an idea about how to balance those requirements?

Regardless, this was a pretty fast-and-dirty approach to get something out the door so I could build other things.  I welcome improvements, or indeed replacement if you come up with something great. :)

Comment (russellpierce) [reply R/awsjavasdk-internal.R:22]: This pattern by the way was borrowed more-or-less from my memory of a circa 2011 package that was part of the ever-frustrating attempts to get XLS data into R.  I am not fond of it either.

Comment (russellpierce) [reply R/awsjavasdk-internal.R:22]: Brainstorming from the above. I'm guessing the most sensible thing would be to publish on the last patch of any given minor release, i.e. when AWS releases the next minor you release the previous build.  That way, presumably the user has the 'bugs worked out' version of that minor release.  If you lock the package version to the AWS SDK version, then you make your `package:AWR` easy to build dependencies on.

Comment (daroczig) [reply R/awsjavasdk-internal.R:22]: Yeah, I was planning to do exactly what you have just suggested: bundling the jar files with the package and pushing to CRAN on every new minor version (like once a year). If someone needs a newer version, then installing from GH/drat should be easy -- so I think this would be a fairly good compromise both to users, CRAN maintainers and R devs.

I'm putting together a build job that can automate this by checking https://github.com/aws/aws-sdk-java/releases for new versions, building the jars and bundling those in a new version of the R package tagged on GH with the AWS version number, so that it'd be easy to install any version of the jars from R.

Will report back here if I have anything useful.

Comment (russellpierce) [reply R/awsjavasdk-internal.R:22]: Definitely let me know when you get it worked out and on CRAN.  Then I'll End-Of-Life this package with a pointer to yours.

Comment (daroczig) [reply R/awsjavasdk-internal.R:22]: Thanks, a lot, @drknexus -- if you have some time, can you please review https://gitlab.com/daroczig/AWR? I tried to push this to GH, but being well above the 1Gb soft repo size limit, I had to pick another repository host provider. Added `drat` for easier installation of the most recent version of the SDK, and the nighly auto-builder is also in progress. Will see if CRAN is willing to accept it :pray:

Comment (russellpierce) [reply R/awsjavasdk-internal.R:22]: @daroczig Nice to see the follow through. =)  

Maybe you already tried this and it didn't work, but `.jpackage("AWR")` should work (and would be more clear).  It seems to me that the default is to load all jars in the `java` directory of the named package.  

You have a typo in the package documentation where the initial a in `JAva` is capitalized.  

I recommend building out a vignette (seems to increase the likelihood of garnering a mention on blogs, and after all if nobody knows about your package, it provides less value to the community).

Other than that, looks good to me.

Comment (daroczig) [reply R/awsjavasdk-internal.R:22]: Thanks, @drknexus! It seems CRAN is not willing to host the ~60Mb of jar files, so ended up creating a wrapper package to be pushed to CRAN which installs the `AWR` package from my private drat repo. Probably it will include some renaming (eg current `AWR` renamed to `AWR.jars` hosted on GitLab and `AWR` to pushed to CRAN) -- will report back later if you are interested in the outcome.

Comment (russellpierce) [reply R/awsjavasdk-internal.R:22]: @daroczig I'm sorry to hear about that re: CRAN.  As I understand it, CRAN also doesn't allow packages which install additional things without intervention by the end-user.  So, I think you're going to have to require the user to run a function to trigger the install of `AWR.jars` from drat.  

At which point, I think you're almost exactly duplicating what `awsjavasdk` already does (but adding a middle-person).  The only major difference I can think of is that, I don't load the `jar`s automatically (when they are present).  However, I'd gladly take a PR to modify .onLoad to make that behavior change and add you as an author.

Comment (daroczig) [reply R/awsjavasdk-internal.R:22]: @drknexus -- AWR is on CRAN after a few iterations: https://cran.r-project.org/package=AWR

Unfortunately, due to the repo design (see below), it was not possible to come up with a PR.

The "trick" (after some conversation with the CRAN maintainers) is that the package automatically downloads and installs the jars from the GitLab repository on first load (which is technically part of the installation), so end-users don't have to manually run a function to download the jars.

The related GitLab repository also stores all AWS SDK versions as per tags and the AWR package can be installed with any AWS SDK version thanks to this design. The only disadvantage of this approach is the pretty big repo size (currently around 6.5 gigs), but fortunately, GitLab allows that (although not possible on GitHub).

The first package using the jars bundled with `AWR` was also published on CRAN in the past few days at https://cran.r-project.org/package=AWR.KMS, and another package on Kinesis is to be pushed there as well after some refactoring.

Comment (russellpierce) [reply R/awsjavasdk-internal.R:22]: Good work with pushing your point with the CRAN maintainers and getting them to change their point of view! I'll give it a look. Congratulations!

Comment (daroczig) [reply R/awsjavasdk-internal.R:22]: Thanks a lot, @drknexus, although I don't think I got them to change their point of view :)
They were extremely helpful with trying to come up with a solution that's convenient for all parties, and after a couple of e-mails this suggestions was accepted. Now the package is very small (only a few kilobytes), so the archive footprint is negligible, although the binary packages on CRAN still include the jar files, so an overall ~100 MB extra storage -- but I hope many R packages will be able to build on the top of that.

BTW do you plan to attend http://unconf17.ropensci.org? I think there will be some R/AWS-related fun, see eg https://twitter.com/sjackman/status/830536912376967169
